gsy_framework/sim_results/carbon_emissions.py

```diff
@@ -0,0 +1,169 @@
+            "periodStart": start_str,
+            "periodEnd": end_str,
+        }
+        params.update(base_params)
+        response = self.session.get(url=URL, params=params, proxies=None, timeout=None)
+        return response
+
+    def _query_generation_per_plant(
+        self, country_code: Union[Area, str], start: pd.Timestamp, end: pd.Timestamp
+    ):
+        """Overwrites EntsoePandasClient query_generation_per_plant method"""
+        area = lookup_area(country_code)
+        params = {
+            "documentType": "A73",
+            "processType": "A16",
+            "in_Domain": area.code,
+        }
+        response = self._base_request(params=params, start=start, end=end)
+        text = response.text
+        df = parse_generation(text, per_plant=True, include_eic=False)
+        df.columns = df.columns.set_levels(
+            df.columns.levels[0].str.encode("latin-1").str.decode("utf-8"), level=0
+        )
+        df = df.tz_convert(area.tz)
+        # Truncation will fail if data is not sorted along the index in rare
+        # cases. Ensure the dataframe is sorted:
+        df = df.sort_index(axis=0)
+
+        if df.columns.nlevels == 2:
+            df = (
+                df.assign(newlevel="Actual Aggregated")
+                .set_index("newlevel", append=True)
+                .unstack("newlevel")
+            )
+        df = df.truncate(before=start, after=end)
+        return df
+
+    def _calculate_total_carbon_generated(
+        self, country_code: str, start: pd.Timestamp, end: pd.Timestamp, stat: str = "median"
+    ) -> pd.DataFrame:
+        """Calculate the total carbon generated based on the specified statistic
+        (min, median, max)"""
+        if stat not in ["min", "median", "max"]:
+            raise ValueError("stat must be one of 'min', 'median', or 'max'")
+
+        df = self._query_generation_per_plant(country_code, start, end)
+
+        generation_types = df.iloc[0].iloc[1:]
+        carbon_emission_per_plant = lambda x: GENERATION_PLANT_TO_CARBON_EMISSIONS[x][
+            stat
+        ]  # noqa: E731 C3001
+        emissions_map = generation_types.map(carbon_emission_per_plant)
+
+        df_numeric = df.iloc[2:].reset_index(drop=True)
+        df_numeric = df_numeric.set_index("Unnamed: 0").apply(  # noqa: E501
+            pd.to_numeric, errors="coerce"
+        )
+        df_numeric.index = pd.to_datetime(df_numeric.index)
+        df_numeric["Total Energy (kWh)"] = df_numeric.sum(axis=1) * 1000  # convert MWh to kWh
+        df_numeric["Total Carbon (gCO2eq)"] = df_numeric.mul(emissions_map, axis=1).sum(axis=1)
+
+        generation_types = df.iloc[0]
+        emissions_map = generation_types.map(
+            lambda x: GENERATION_PLANT_TO_CARBON_EMISSIONS.get(x, {}).get(stat, 0)
+        )
+        return df_numeric[["Total Energy (kWh)", "Total Carbon (gCO2eq)"]]
+
+    def update(self, country_code: str, time_slot: duration, results_list: List[Dict]):
```

Comment:
With putting the country_code in the interface of the method, you are assuming that all areas in the simulation are in the same country. Works for now, but how do you get the country code ? Curious about your gsy-web and gsy-e PRs.

Reply:
I thought it would be best to accept the country code rather than geo coordinates, because it would simplify when we use it in research-scripts for example, but now that I think it further maybe it is best to calculate the country code here

Reply:
Yes, we need the location of the area in any case and need to pass it here:  `ResultsCalculation._calculate_current_market_sim_results`